### DISCUSSION TITLE
Pixel Shader support

### DISCUSSION THREAD

**Discussion by ivoryguard** [AvaloniaUI/Avalonia#10388] [Q&A] [answered]
Hello.

I tried Googling whether pixel shaders are supported in AvaloniaUI, but I couldn't find an exact answer.

My existing WPF app uses Pixel Shader effect using HLSL (.ps file).
So, I need pixel shader effect in AvaloniaUI in order to port existing WPF apps to AvaloniaUI on Windows and Android.

I suspect that it is possible at AvaloniUI theoretically, because it uses SkiaSharp for pixel perfect rendering.

Can I use pixel shader effects simliar to WPF at AvaloniaUI?

Best regards.

**Comment by maxkatz6** [comment, marked as answer]
> Can I use pixel shader effects simliar to WPF at AvaloniaUI?

Not in Avalonia. But you can use SkSL by accessing SkiaSharp directly.
I.e. a combination of https://gist.github.com/kekekeks/ac06098a74fe87d49a9ff9ea37fa67bc and https://github.com/mono/SkiaSharp/issues/1319#issuecomment-640529824

> So, I need pixel shader effect in AvaloniaUI in order to port existing WPF apps to AvaloniaUI on Windows and Android.

There is a problem with defining well supported shaders format.
HLSL is DirectX only, so Windows only. Technically Vulkan also supports it now.
GLSL is OpenGL only. Might have problems on metal ios/macos or directx windows, but still the best choice probably.
SKSL is Skia specific, but also has unsable API, as Google doesn't maintain skia API stability.

**Reply by ivoryguard** [reply]
Thank you for the reply.

What I want really is color inversion of a specific UserControl or ContentControl.

I saw the source code in the link. It is interesting.
I have some experience of HLSL and OpenCVSharp, and so it can be possible to create my own color filter like SkSL.

Is SkSL fast enough like the effect function of WPF?
